### PR TITLE
EPMDJ-2523: Сreate a component for Filtering with checkboxes

### PR DESCRIPTION
**Pull request by RomanAchilov**
Issue: EPMDJ-2523

### CODE REVIEW

File: src/components/filtering-with-checkboxes/filter-with-checkboxes.tsx [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import styles from './filter-with-checkboxes.module.scss';
+
+interface Props {
+  className?: string;
+  filters: string[];
+}
+
+const filterWithCheckboxes = BEM(styles);
+
+export const FilterWithCheckboxes = filterWithCheckboxes(({ className, filters }: Props) => {
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/filter-with-checkboxes.tsx]: TableFilter

File: src/components/filtering-with-checkboxes/filter-with-checkboxes.tsx [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  const {
+    filterItems, selectedItems, handleChange, handleAllChange,
+  } = useFilter(filters, isExpanded);
+
+  return (
+    <div className={className} ref={node}>
+      <div onClick={() => setIsExpanded(!isExpanded)}>
+        {selectedItems.length > 0 && selectedItems.length < filters.length ? selectedItems.map(({ name }) => (
+          <SelectedItems key={nanoid()}>
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/filter-with-checkboxes.tsx]: name can be used as key

File: src/components/filtering-with-checkboxes/use-filter.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { useState, useEffect } from 'react';
+
+export const useFilter = (filters: string[], isExpanded: boolean) => {
+  const [filterItems, setFilterItems] = useState(filters.map(filter => {
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/use-filter.ts]: i dont even understand why u store all items at store? u can store only selected items

File: src/components/filtering-with-checkboxes/use-filter.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    const filterItem = { name: filter, isChecked: true };
+    return filterItem;
+  }));
+
+  const selectedItems = filterItems.filter((filterItem) => filterItem.isChecked && filterItem);
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/use-filter.ts]: `filterItem.isChecked && filterItem` what is this?

Comment (pavyarov) [reply src/components/filtering-with-checkboxes/use-filter.ts]: i mean if `filterItem` can be null or undefined... ect - this call `filterItem.isChecked` will trigger TypeError

File: src/components/filtering-with-checkboxes/use-filter.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { name, checked } = event.target;
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/use-filter.ts]:
```suggestion
  const handleChange = ( target:  { name, checked } }: React.ChangeEvent<HTMLInputElement>) => {
```

File: src/components/filtering-with-checkboxes/use-filter.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    setFilterItems(filterItems.map((filterItem) => {
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/use-filter.ts]: this is `.reduce()`

File: src/components/filtering-with-checkboxes/use-filter.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  const handleAllChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const { checked } = event.target;
+    setFilterItems(filterItems.map((filterItem) => {
+      const modifiedFilterItem = { ...filterItem };
+      modifiedFilterItem.isChecked = checked;
+      return modifiedFilterItem;
+    }));
+  };
+
+  useEffect(() => {
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/use-filter.ts]: for what reason u whote this effect?

File: src/components/filtering-with-checkboxes/use-filter.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import { useState, useEffect } from 'react';
+
+export const useFilter = (filters: string[], isExpanded: boolean) => {
```
Comment (pavyarov) [inline src/components/filtering-with-checkboxes/use-filter.ts]: i think you should use useReducer there

File: src/components/filtering-with-checkboxes/use-filter.ts (hunk 7 above)
Comment (RomanAchilov) [reply src/components/filtering-with-checkboxes/use-filter.ts]: > for what reason u whote this effect?

If User unchecks all checkboxes or User clicks on "All" checkbox and then User closes "Filter" list and User opens it again, all checkboxes are CHECKED

File: src/components/table-filter/use-filter.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    case TOGGLE_ALL_CHECKBOXES:
+      return checkboxes.map(checkbox => ({ ...checkbox, checked: !checkbox.checked }));
+    default:
+      return checkboxes;
+  }
+};
+
+export const useFilter = (filters: string[]) => {
+  const [isExpanded, setIsExpanded] = useState(false);
+  const [checkboxes, dispatch] = useReducer(reducer, filters.map(filter => ({ name: filter, checked: true })));
```
Comment (pavyarov) [inline src/components/table-filter/use-filter.ts]: why state is not properly typed

File: src/components/table-filter/use-filter.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    dispatch(toggleAllCheckboxes(checkboxes));
+  };
+
+  useEffect(() => {
+    checkboxes.every(checkbox => checkbox.checked === false) &&
+    handleToggleAllCheckboxes();
+  }, [isExpanded]);
+
+  return {
+    isExpanded, setIsExpanded, checkboxes, selectedItems, handleToggleCheckbox, handleToggleAllCheckboxes,
```
Comment (pavyarov) [inline src/components/table-filter/use-filter.ts]: isExpanded, dispatch, selectedItems

**Pull request closed without merging**